diff --git a/Makefile b/Makefile
index 2380e79..98f2bd0 100644
--- a/Makefile
+++ b/Makefile
@@ -68,10 +68,6 @@ install:	minissdpd
 	$(INSTALL) minissdpd $(SBININSTALLDIR)
 	$(INSTALL) -d $(MANINSTALLDIR)/man1
 	$(INSTALL) minissdpd.1 $(MANINSTALLDIR)/man1/minissdpd.1
-ifeq (, $(findstring darwin, $(OS)))
-	$(INSTALL) -d $(PREFIX)/etc/init.d
-	$(INSTALL) minissdpd.init.d.script $(PREFIX)/etc/init.d/minissdpd
-endif
 
 check:	validateminissdpd validatecodelength
